jwst/model_blender/blender.py

```diff
@@ -1,254 +1,152 @@
-    for model in input_models:
-        if not isinstance(model, JwstDataModel):
-            if not isinstance(model, str):
-                raise TypeError(
-                    "Each entry in the headers list must be either a " +
-                    "datamodels.JwstDataModel instance or a filename (str)")
-            model = datamodels.open(model)
-        header = model.to_flat_dict()
-        filename = header['meta.filename']
-        for i, mapping in enumerate(mappings):
-            if mapping.src_kwd in header:
-                value = model[mapping.src_kwd]
-            elif mapping.error_type == 'raise':
+    def __init__(self, blend_ignore_attrs=None):
+        """
+        Create a new `ModelBlender` using an optional list of
+        attributes to ignore.
+
+        Parameters
+        ----------
+        blend_ignore_attrs: list or None
+            A list of metadata attributes to ignore during blending.
+            These attributes will not be set on the output/combined.
+            These attributes must be strings containing the dotted
+            path of each attribute (for example "meta.filename").
+            (Note that "meta.wcs" will always be ignored).
+        """
+        self._model_type = None
+        self._first_header_meta = None
+        self._blenders = None
+        self._table_builder = None
+        self._blend_ignore_attrs = ['meta.wcs']
+        if blend_ignore_attrs is not None:
+            self._blend_ignore_attrs.extend(blend_ignore_attrs)
+
+    def accumulate(self, model):
+        """
+        Process a datamodel, adding it's metadata to the blended
+        metadata and the metadata table.
+
+        Parameters
+        ----------
+
+        model: `jwst.datamodels.JwstDataModel`
+            The datatamodel to blend.
+        """
+        if self._first_header_meta is None:
+            self._model_type = type(model)
+            # search the schema for other metadata to "blend" and to add to the table
+            attr_to_columns, attr_to_blend_rules, schema_ignores  = parse_schema(model.schema)
+
+            # update ignores list for items in schema that can't be blended
+            self._blend_ignore_attrs.extend(schema_ignores)
+
+            # capture the entire contents of the first model metadata
+            self._first_header_meta = {
+                attr: v
+                for attr, v in model.to_flat_dict(include_arrays=False).items()
+                if attr.startswith('meta') and not any((attr.startswith(i) for i in self._blend_ignore_attrs))
+            }
+
+            # make "blenders" for the metadata with special rules
+            self._blenders = {
+                attr: make_blender(rule)
+                for attr, rule in attr_to_blend_rules.items()
+                # 'first' is the default behavior so these rules need not be made explicitly
+                if rule != 'first' and not any((attr.startswith(i) for i in self._blend_ignore_attrs))
+            }
+
```

Comment:
This is a style choice so your call, but I think these would be more readable/maintainable if they were written as for loops instead of comprehensions.

Reply:
My brain likes that much better.  Thanks!